src/control/destination/background/mod.rs

```diff
@@ -317,15 +353,42 @@ where
 // ===== impl Config =====
 
 impl Config {
-    /// Initiates a query `query` to the Destination service and returns it as
-    /// `Some(query)` if the given authority's host is of a form suitable for using to
-    /// query the Destination service. Otherwise, returns `None`.
+
+    fn new(namespaces: Namespaces, max_queries: usize) -> Self {
+        Config {
+            namespaces,
+            max_queries,
+            active_queries: Arc::new(()),
+        }
+    }
+
+    /// Returns true if there is currently capacity for additional
+    /// Destination service queries.
+    fn has_more_queries(&self) -> bool {
+        Arc::weak_count(&self.active_queries) < self.max_queries
+    }
+
+    /// Attepts to initiate a query `query` to the Destination service
+    /// if the given authority's host is of a form suitable for using to
+    /// query the Destination service.
+    ///
+    /// # Returns
+    ///
+    /// - `(None, true)` if the given authority _should_ query the
+    ///   Destination service, but the query limit has already been reached.
+    ///   In this case, this function should be called again when there is
+    ///   capacity for additional queries.
+    /// - `(None, false)` if the authority is not suitable for querying the
+    ///   Destination service.
+    /// - `(Some(DestinationServiceQuery), true)` if the authority is suitable
+    ///   for querying the Destination service and there is sufficient capacity
+    ///   to initiate a new query.
     fn query_destination_service_if_relevant<T>(
         &self,
-        client: &mut T,
+        client: Option<&mut T>,
         auth: &DnsNameAndPort,
         connect_or_reconnect: &str,
-    ) -> Option<DestinationServiceQuery<T>>
+    ) -> (Option<DestinationServiceQuery<T>>, bool)
```

Comment:
One obvious criticism is that it seems more idiomatic to return a `Result<Option<DestinationServiceQuery<T>>, TooManyErrors>` here. That was my original feeling as well. 

However, when I did so, it ended up being fairly repetitive, as every time this function was called, we had to unpack the result into an `Option` and a `bool` _anyway_. Just having this return `Option<...>, bool)` obviates the need for repeating that code.